fix(error): adding type to error in string validator in introspect

santanu8961 opened the pull request:
<!--- Provide a formatted commit message describing this PR in the Title above -->
<!--- See our DEVELOPERS guide below: -->
<!--- https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format -->
# Closes <https://github.com/accordproject/concerto/issues/772>
<!--- Provide an overall summary of the pull request -->

### Changes
<!--- More detailed and granular description of changes -->
Adds type to validator error in introspect validator.
<!--- These should likely be gathered from commit message summaries -->
- added a new optional parameter to class `BaseException`  which can take `errorType` as well as `errorMessage` which helps to determine what kind of error is getting thrown from the validator itself. and using that to throw error in `validator`.
- also added this in `NotFoundException` for better complience


### Author Checklist
- [x] Ensure you provide a [DCO sign-off](https://github.com/probot/dco#how-it-works) for your commits using the `--signoff` option of git commit.
- [ ] Vital features and changes captured in unit and/or integration tests
- [x] Commits messages follow [AP format](https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format)
- [ ] Extend the documentation, if necessary
- [x] Merging to `main` from `fork:branchname`

sanketshevkar commented on packages/concerto-core/lib/introspect/stringvalidatorexception.js:
```diff
@@ -0,0 +1,38 @@
+'use strict';
+
+/**
+* Class representing an validation error related to anykind of string validatoin in introspect
```
nit: *validatoin

sanketshevkar commented on packages/concerto-core/lib/introspect/stringvalidatorexception.js:
```diff
@@ -0,0 +1,38 @@
+    /**
+     * Create error related to string validation.
+     * @param {string} errormessage - The exception message.
+     * @param {string} errorType - conditional parameter which denoted the kind of string
```
nit: *denoted -> denotes/represents

sanketshevkar commented:
Should we have a new folder for error classes and not keep it in introspect?

mttrbrts commented on packages/concerto-core/lib/introspect/stringvalidatorexception.js:
```diff
@@ -0,0 +1,38 @@
+* @class
+* @memberof module:concerto-core
+*/
+class StringValidationException extends Error {
```
It should be Validator, not Validation. Internally the `regex` keyword is called a `StringValidator`.

dselman replied:
Yes, we should ensure we can use similar approach with NumberValidator as well.

santanu8961 replied:
Hi @mttrbrts , @ekarademir  I am keeping the name of the class as `StringValidatiorException` and the default error as `ValidatorError` and the regex error as `RegexValidatorException`.
let me know if this works fine . and this is what you meant.

dselman commented on packages/concerto-core/lib/introspect/validator.js:
```diff
@@ -46,10 +48,11 @@ class Validator {
      */
-    reportError(id, msg) {
-        throw new Error( 'Validator error for field `' + id + '`. ' + this.getFieldOrScalarDeclaration().getFullyQualifiedName() + ': ' + msg );
+    reportError(id, msg, errorType='ValidatorError') {
```
Let's avoid having to use `instanceof` checks - those make assumptions about how packages are required (instanceof can fail if the same code is loaded at different version) which have bitten us in the past.

santanu8961 replied:
Hi @ekarademir I have used the additional argument `errorType` so that it can be propagated through till `StringValidationException` where we can use it to denote what kind of error it is , which can be very useful to do any kind of operation as only error string is quick tricky to detect what error it is trying to throw

santanu8961 replied:
Hi @dselman , can you please clarify what do you mean by the instance of checks?
Thanks.

dselman replied:
I mean client code doing things like:

```
try {
...
}
catch(err) {
if(err instanceof StringValidatorException) {
 ...
}
}
```

dselman reviewed (changes requested):
I like the motivation for this. I just needs a bit of refactoring to ensure it works with NumberValidator and can be cleanly consumed upstream.

dselman commented on packages/concerto-core/lib/introspect/stringvalidator.js:
```diff
@@ -70,7 +70,7 @@ class StringValidator extends Validator{
             }
             catch(exception) {
-                this.reportError(field.getName(), exception.message);
+                this.reportError(field.getName(), exception.message,'RegexValidatorException');
```
I think if we want to go this route then the error codes should be defined and exported and referenced.

E.g. something like:

```
ErrorCodes.js
export const INVALID_REGEX = 1;
```

Somewhere we want to report an error:
```
this.reportError(field.getName(), exception.message, ErrorCodes.INVALID_REGEX);
```

This allows a client to do something like:

```
import {ErrorCodes} from '@accordproject/concerto-core';

try {
...
}
catch(err) {
const errorCode = err?.getErrorCode.?();
switch(errorCode) {
case ErrorCodes.INVALID_REGEX:
  // do something specific here
break;
default:
  // general case
}
```

We should have in mind that concerto-core will be moving to TypeScript and we want these error codes to be type safe.

santanu8961 replied:
Hi @dselman , as you requested I have changed the structure by doing the following changes to the structure:
1. I no longer am using the StringValidatorException class instead I have added the errorType as an optional param in BaseException and using that.
2. added an errorCodes.js file where I am storing the error type constants and using them directly , these are exposed so the client will also be able to use these , as discussed in the PR

dselman commented on packages/concerto-core/lib/introspect/stringvalidatorexception.js:
```diff
@@ -0,0 +1,38 @@
+* @class
+* @memberof module:concerto-core
+*/
+class StringValidatorException extends Error {
```
This could be more generic - it extends Error and adds an Error Code. We could add an optional error code to concerto-util/BaseException, so that all exceptions can reference an error code.

dselman reviewed (changes requested):
I would prefer to see generic Error Code functionality moved into concerto-util/BaseException, and to refactor code to throw Errors that derive from BaseException.

Error codes should be defined constants so that code is type safe.

santanu8961 commented on packages/concerto-core/lib/introspect/stringvalidator.js:
```diff
@@ -70,7 +71,7 @@ class StringValidator extends Validator{
             }
             catch(exception) {
-                this.reportError(field.getName(), exception.message);
+                this.reportError(field.getName(), exception.message,ErrorCodes.REGEX_VALIDATOR_EXCEPTION);
```
Here I have used error codes as constants instead of inline strings for consistency

santanu8961 commented on packages/concerto-core/lib/introspect/validator.js:
```diff
@@ -46,10 +47,11 @@ class Validator {
      */
-    reportError(id, msg) {
-        throw new Error( 'Validator error for field `' + id + '`. ' + this.getFieldOrScalarDeclaration().getFullyQualifiedName() + ': ' + msg );
+    reportError(id, msg, errorType=ErrorCodes.DEFAULT_VALIDATOR_EXCEPTION) {
```
Here I have used error codes as constants instead of inline strings for consistency

santanu8961 commented on packages/concerto-core/lib/introspect/validator.js:
```diff
@@ -46,10 +47,11 @@ class Validator {
-    reportError(id, msg) {
-        throw new Error( 'Validator error for field `' + id + '`. ' + this.getFieldOrScalarDeclaration().getFullyQualifiedName() + ': ' + msg );
+    reportError(id, msg, errorType=ErrorCodes.DEFAULT_VALIDATOR_EXCEPTION) {
+        throw new BaseException('Validator error for field `' + id + '`. ' + this.getFieldOrScalarDeclaration().getFullyQualifiedName() + ': ' + msg ,errorType);
```
removed StringValidatorExceptionClass as its not needed anymore I have added a optional param errorType in BaseException which solves the purpose

santanu8961 commented on packages/concerto-util/lib/errorcodes.js:
```diff
@@ -0,0 +1,22 @@
+ */
+
+'use strict';
+
```
error codes are defined in this file which can be used to implement different errors when ever needed

santanu8961 commented on packages/concerto-util/lib/baseexception.js:
```diff
@@ -27,12 +30,14 @@ class BaseException extends Error {
+     * @param {string} errorType - The optional error code regarding the error
      */
-    constructor(message, component) {
+    constructor(message, component, errorType) {
```
added a errortype parameter which is optional and provides the errorcode related to the error

santanu8961 commented:
Hi @dselman , as you requested I have changed the structure by doing the following changes to the structure:
1. I no longer am using the StringValidatorException class instead I have added the errorType as an optional param in BaseException and using that.
2. added an errorCodes.js file where I am storing the error type constants and using them directly , these are exposed so the client will also be able to use these , as discussed in the PR .

requesting your approval on this PR

sanketshevkar commented on packages/concerto-core/lib/introspect/stringvalidator.js:
```diff
@@ -70,7 +71,7 @@ class StringValidator extends Validator{
             }
             catch(exception) {
-                this.reportError(field.getName(), exception.message);
+                this.reportError(field.getName(), exception.message,ErrorCodes.REGEX_VALIDATOR_EXCEPTION);
```
nit: lint check in pipeline could throw error here <spacing issue>

santanu8961 replied:
resolved

sanketshevkar commented on packages/concerto-core/types/lib/typenotfoundexception.d.ts:
```diff
@@ -7,14 +7,6 @@ export = TypeNotFoundException;
-     * @param {string|undefined} message - error message.
-     * @param {string} component - the optional component which throws this error
-     */
-    constructor(typeName: string, message: string | undefined, component: string);
```
On the main I still see constructor on `TypeNotFoundException.js` but why is this getting deleted here? Is this auto generated?

santanu8961 replied:
Yes this is actually auto generated after I run build command

sanketshevkar commented on packages/concerto-core/lib/introspect/validator.js:
```diff
@@ -46,10 +47,11 @@ class Validator {
-    reportError(id, msg) {
-        throw new Error( 'Validator error for field `' + id + '`. ' + this.getFieldOrScalarDeclaration().getFullyQualifiedName() + ': ' + msg );
+    reportError(id, msg, errorType=ErrorCodes.DEFAULT_VALIDATOR_EXCEPTION) {
+        throw new BaseException('Validator error for field `' + id + '`. ' + this.getFieldOrScalarDeclaration().getFullyQualifiedName() + ': ' + msg ,undefined,errorType);
```
nit: Space linting, same as above.

santanu8961 replied:
resolved

dselman reviewed (approved):
Good work. Thanks!
